webapp/src/auth.ts

```diff
@@ -159,46 +155,92 @@ export function identityProviders(): pxt.AppCloudProvider[] {
         .filter(prov => prov.identity);
 }
 
+export function hasIdentity(): boolean {
+    return identityProviders().length > 0;
+}
+
+export function loggedIn(): boolean {
+    if (!hasIdentity()) return false;
+    const state = getState();
+    return !!state.user?.id
+}
+
+export function profileNeedsSetup(): boolean {
+    const state = getState();
+    return loggedIn() && !state.user.username;
+}
+
+export async function updateUserProfile(opts: {
+    username?: string,
+    avatarUrl?: string
+}) {
+    if (!loggedIn()) return;
+    const result = await apiAsync<UserProfile>('/api/user/profile', {
+        username: opts.username,
+        avatarUrl: opts.avatarUrl
+    } as UserProfile);
+    if (result.success) {
+        // Set user profile from returned value
+        setUser(result.resp);
+    }
+}
+
 /**
  * Private functions
  */
 
-const redirBase = () => `${window.location.protocol}//${window.location.host}`;
-
 async function fetchUserAsync() {
     const state = getState();
-    if (state.user) return;
 
-    const resp = await apiAsync('user/profile');
-    const user: UserProfile = resp.json;
-
-    setUser(user);
-}
-
-function mkLoginStateKey(id: string): string {
-    return `auth.login.state:${id}`;
-}
+    // We already have a user, no need to get it again.
+    if (state.user) return;
```

Comment:
What's the expiration on this?

Reply:
On `state.user`? It lives in memory, never stored locally.